hudi-spark-datasource/hudi-spark/src/test/java/org/apache/hudi/TestMetadataTableSupport.java

```diff
@@ -0,0 +1,99 @@
+import org.apache.hudi.common.table.timeline.HoodieActiveTimeline;
+import org.apache.hudi.common.table.timeline.HoodieInstant;
+import org.apache.hudi.config.HoodieCompactionConfig;
+import org.apache.hudi.config.HoodieWriteConfig;
+import org.apache.hudi.metadata.HoodieTableMetadata;
+import org.apache.hudi.storage.StoragePath;
+import org.apache.hudi.testutils.HoodieSparkClientTestBase;
+
+import org.apache.spark.api.java.JavaRDD;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import java.util.List;
+
+import static org.apache.hudi.common.table.timeline.HoodieTimeline.REPLACE_COMMIT_ACTION;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+class TestMetadataTableSupport extends HoodieSparkClientTestBase {
+  @BeforeEach
+  void start() throws Exception {
+    super.setUp();
+  }
+
+  @AfterEach
+  void end() throws Exception {
+    super.tearDown();
+  }
+
+  @Test
+  void testRecreateMDTForInsertOverwriteTableOperation() {
+    HoodieWriteConfig config = getConfigBuilder()
+        .withCompactionConfig(HoodieCompactionConfig.newBuilder()
+            .withMaxNumDeltaCommitsBeforeCompaction(1).build())
+        .withMetadataConfig(HoodieMetadataConfig.newBuilder()
+            .withEnableRecordIndex(true).build())
+        .build();
+
+    try (SparkRDDWriteClient writeClient = getHoodieWriteClient(config)) {
+      // Insert first batch.
+      String timestamp0 = "20241015000000000";
+      List<HoodieRecord> records0 = dataGen.generateInserts(timestamp0, 100);
+      JavaRDD<HoodieRecord> dataset0 = jsc.parallelize(records0, 2);
+
+      writeClient.startCommitWithTime(timestamp0);
+      writeClient.insert(dataset0, timestamp0).collect();
+
+      // Insert second batch.
+      String timestamp1 = "20241015000000001";
+      List<HoodieRecord> records1 = dataGen.generateInserts(timestamp1, 50);
+      JavaRDD<HoodieRecord> dataset1 = jsc.parallelize(records1, 2);
+
+      writeClient.startCommitWithTime(timestamp1, REPLACE_COMMIT_ACTION);
+      writeClient.insertOverwriteTable(dataset1, timestamp1);
+
+      // Validate.
+      StoragePath mdtBasePath =
+          HoodieTableMetadata.getMetadataTableBasePath(metaClient.getBasePath());
+      HoodieTableMetaClient mdtMetaClient = HoodieTableMetaClient.builder()
+          .setConf(storageConf.newInstance())
+          .setBasePath(mdtBasePath).build();
+
+      HoodieActiveTimeline timeline = mdtMetaClient.getActiveTimeline();
+      List<HoodieInstant> instants = timeline.getInstants();
+      assertEquals(3, timeline.getInstants().size());
+      // For MDT bootstrap instant.
+      assertEquals("00000000000000000", instants.get(0).getTimestamp());
+      // For RLI bootstrap instant.
+      assertEquals("00000000000000001", instants.get(1).getTimestamp());
+      // For the insert_overwrite_table instant.
+      assertEquals(timestamp1, instants.get(2).getTimestamp());
+    }
+  }
```

Comment:
in the end, we can validate that one of the commits in MDT (may be latest completed) before triggering insert overwrite operation is not present anymore in the end (after insert overwrite)

Reply:
Checked the first insert commit exists before the second insert.